[ASan][Compiler-rt] Support Gracefull hsa-runtime shutdown.

ampandey-AMD opened the pull request:
Summary:
  - Track runtime shutdown via AmdgpuMemFuncs::IsAmdgpuRuntimeShutdown()
    and gate further hsa_amd_pointer_info calls.
  - Add interception of 'hsa_init' api call.
  - Add registration of hsa-runtime associated system events via
    'hsa_amd_register_system_event_handler'.
    - HSA event registered is 'HSA_AMD_SYSTEM_SHUTDOWN_EVENT'

ampandey-AMD committed:
[ASan][Compiler-rt] Support Gracefull hsa-runtime shutdown.

Summary:
  - Track runtime shutdown via AmdgpuMemFuncs::IsAmdgpuRuntimeShutdown()
    and gate further hsa_amd_pointer_info calls.
  - Add interception of 'hsa_init' api call.
  - Add registration of hsa-runtime associated system events via
    'hsa_amd_register_system_event_handler'.
    - HSA event registered is 'HSA_AMD_SYSTEM_SHUTDOWN_EVENT'

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2765

ampandey-AMD committed:
Fix clang-format noise.

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2889

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2943

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2944

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2945

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2946

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2984

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/2996

ampandey-AMD committed:
Refactor atomic variable checking/updation.

Summary:
  - Added wrapper IsAmdgpuRuntimeShutdown for checking state.
  - Added wrapper NotifyAmdgpuRuntimeShutdown for updating state.

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/3002

b-sumner commented on compiler-rt/lib/sanitizer_common/sanitizer_allocator_amdgpu.cpp:
```diff
@@ -26,14 +28,36 @@ struct HsaMemoryFunctions {
+// Notify AMDGPU runtime shutdown to allocator
+void AmdgpuMemFuncs::NotifyAmdgpuRuntimeShutdown() {
+  uint8_t shutdown = 0;
+  if (atomic_compare_exchange_strong(&amdgpu_runtime_shutdown, &shutdown, 1,
```
Wouldn't atomic fetch-OR work, or is there a reason to prefer CAS?

b-sumner replied:
Oops.  OR is not available for some reason.

b-sumner commented on compiler-rt/lib/sanitizer_common/sanitizer_allocator_amdgpu.cpp:
```diff
@@ -102,6 +135,28 @@ bool AmdgpuMemFuncs::GetPointerInfo(uptr ptr, DevicePointerInfo* ptr_info) {
+ // TODO: Register multiple event handlers if needed in future
+void AmdgpuMemFuncs::RegisterSystemEventHandlers() {
+  // Check if already registered
+  if (atomic_load(&amdgpu_event_registered, memory_order_acquire) == 0) {
```
Does this need to be a CAS to prevent multiple threads from registering the callback?

ampandey-AMD replied:
Yes, thanks multiple callbacks can be registered via hsa_amd_register_system_event_handler for single HSA_AMD_SHUTDOWN_EVENT.

Commit [894289b6](https://github.com/ROCm/llvm-project/pull/559/commits/894289b69a287618a790331ba37dfd9061892f22) added CAS check.

b-sumner commented on compiler-rt/lib/sanitizer_common/sanitizer_allocator_amdgpu.cpp:
```diff
@@ -26,14 +28,36 @@ struct HsaMemoryFunctions {
+  }
+}
+
 bool AmdgpuMemFuncs::Init() {
```
Should we change the name here since it's not only initializing memory functions?

ampandey-AMD replied:
Not sure. @bing-ma can comment more?

I refactored the Init function for the ease of readability and maintenance.

b-sumner replied:
I don't think it's a big concern.  I'll resolve for now.

bing-ma replied:
Agreed it is not a big concern. Still I am not sure if shutdown handling should be here in a common allocator. Unfortunately I don't have time to research the possibility of moving this to somewhere else. I just feel it is strange to handle runtime functionality in an allocator.

ampandey-AMD committed:
Refactor sanitizer_allocator_amdgpu.cpp

  - Add verbose logging logic.
  - Refactor AmdgpuMemFuncs::Init function.
  - Add CAS check for AMDGPU shutdown callback registering.

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/3046

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/3047

b-sumner commented on compiler-rt/lib/sanitizer_common/sanitizer_allocator_amdgpu.cpp:
```diff
@@ -102,6 +161,38 @@ bool AmdgpuMemFuncs::GetPointerInfo(uptr ptr, DevicePointerInfo* ptr_info) {
+      VReport(1,
+              "Amdgpu RegisterSystemEventHandlers: Failed to register shutdown "
+              "event \n");
+      atomic_store(&amdgpu_event_registered, 0, memory_order_release);
```
I guess this doesn't hurt.  But do we really expect that if registration fails for one call of hsa_init it will actually succeed for another call?

ampandey-AMD replied:
Yes, @b-sumner as I recall multiple hsa_init(only first call does initilization rest make a reference count and return HSA_STATUS_SUCCESS) calls are allowed. This function `RegisterSystemEventHandlers` is always executed when hsa_init intercepted by ASan does hsa_init call and is successful. ASan must register a single shutdown callback , however multiple callbacks can be registered for shutdown event, but ASan has to only keep a single one and not waste resources by issuing multiple callbacks(CAS check at line 169) . Also , if  `hsa_amd_register_system_event_handler` call (at line number 182) fails (callback not registered) then another hsa_init call may succeed in registering same callback at later time(line 192 allows that).

b-sumner replied:
Right.  I'm just pointing out that something is extremely wrong if, immediately after a successful call to hsa_init, a registration failure happens.  All that registration is really doing is appending to a list.  And given how bad the situation is to see that failure, it is also unlikely that anything is going to come along and fix things and somehow allow the registration to succeed later on.

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/3061

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/3062

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/3063

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/3086

b-sumner commented on compiler-rt/lib/sanitizer_common/sanitizer_allocator_amdgpu.cpp:
```diff
@@ -26,36 +28,76 @@ struct HsaMemoryFunctions {
 void *AmdgpuMemFuncs::Allocate(uptr size, uptr alignment,
                                DeviceAllocationInfo *da_info) {
+  // Do not allocate if AMDGPU runtime is shutdown
+  if (IsAmdgpuRuntimeShutdown()) {
```
Should we use UNLIKELY here?

b-sumner commented on compiler-rt/lib/sanitizer_common/sanitizer_allocator_amdgpu.cpp:
```diff
@@ -73,6 +115,15 @@ void *AmdgpuMemFuncs::Allocate(uptr size, uptr alignment,
 
 void AmdgpuMemFuncs::Deallocate(void *p) {
+  // Deallocate does nothing after AMDGPU runtime shutdown
+  if (IsAmdgpuRuntimeShutdown()) {
```
Should we use UNLIKELY here?

b-sumner commented on compiler-rt/lib/sanitizer_common/sanitizer_allocator_amdgpu.cpp:
```diff
@@ -85,6 +136,14 @@ void AmdgpuMemFuncs::Deallocate(void *p) {
 
 bool AmdgpuMemFuncs::GetPointerInfo(uptr ptr, DevicePointerInfo* ptr_info) {
+  // GetPointerInfo returns false after AMDGPU runtime shutdown
+  if (IsAmdgpuRuntimeShutdown()) {
```
Should we use UNLIKELY here?

b-sumner commented on compiler-rt/lib/sanitizer_common/sanitizer_allocator_device.h:
```diff
@@ -122,7 +122,8 @@ class DeviceAllocatorT {
       CHECK_LT(idx, n_chunks_);
       h = GetHeader(chunks_[idx], &header);
-      CHECK(!dev_runtime_unloaded_);
+      if (dev_runtime_unloaded_)
```
Should we use UNLIKELY here?

b-sumner commented on compiler-rt/lib/sanitizer_common/sanitizer_allocator_device.h:
```diff
@@ -140,7 +141,8 @@ class DeviceAllocatorT {
     for (uptr i = 0; i < n_chunks_; i++) {
       Header *h = GetHeader(chunks_[i], &header);
-      CHECK(!dev_runtime_unloaded_);
+      if (dev_runtime_unloaded_)
```
Should we use UNLIKELY here?

b-sumner commented on compiler-rt/lib/sanitizer_common/sanitizer_allocator_device.h:
```diff
@@ -306,14 +307,21 @@ class DeviceAllocatorT {
+    // Device allocator has dependency on device runtime. If device runtime
+    // is unloaded, GetPointerInfo() will fail. For such case, we can still
+    // return a valid value for map_beg, map_size will be limited to one page
+    if (!dev_runtime_unloaded_) {
```
Should we instead make an UNLIKELY check for unloaded here?

ampandey-AMD replied:
It should be LIKELY here as this condition would probably be true most of the times.

ampandey-AMD committed:
Add LIKELY, UNLIKEY macros for branch prediction hints for compiler.

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/3174

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/3175

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/3176

b-sumner commented on compiler-rt/lib/sanitizer_common/sanitizer_allocator_device.h:
```diff
@@ -122,7 +122,8 @@ class DeviceAllocatorT {
       CHECK_LT(idx, n_chunks_);
       h = GetHeader(chunks_[idx], &header);
-      CHECK(!dev_runtime_unloaded_);
+      if (UNLIKELY(dev_runtime_unloaded_))
```
Can we drop this test completely now?

ampandey-AMD replied:
I didn't understand. Why drop it?

b-sumner replied:
I think it was there before to avoid a crash or other problem.  But with the other changes in this patch, I don't see a reason to return here.  But maybe I missed something?

ampandey-AMD replied:
From my understanding we shouldn't drop the early check in `DeviceAllocator Deallocate` because in line 129-132 we are amending the data structures by adding the allocation to the n_frees list. Dropping this would cause inconsistent state because `AmdgpuMemFuncs::Deallocate` can early return if shutdown flag is set but the allocation is actually not freed. 

The early guards is necessary for both **Deallocate** functions although they are at different layers i.e one is at `DeviceAllocator` other one at `AmdgpuAllocator`.

b-sumner replied:
I don't think the chunks or stats update will cause any harm...and might actually be good.  I don't think the unmap callback would be a problem, and I think the DeviceMemFuncs::Deallocate is protected.  But I'm not certain.

ampandey-AMD replied:
If we remove this check here then there will be chances of being in an inconsistent state in the `DeviceAllocator` layer which is supposed to be target agnostic from the `AmdgpuAllocator` layer. Previously, in this place we had a hard `CHECK(!dev_runtime_unloaded_)` here whose initialization was actually dependent on `GetPointerInfo(doing device specific **hsa_amd_pointer_info** for any pointer at DeviceAllocator layer)` to switch the value of dev_runtime_unloaded_ to true so as to determine the runtime shutdown state.

We now have a deterministic and sophisticated way of querying the hsa-runtime shutdown state via callback executed at `hsa_shut_down` in  a separate rocr thread. `dev_runtime_unloaded_` is initialized & checked atomically for any pointer deallocation, so it is not dependent now on figuring shutdown state via failure of return status of `hsa_amd_pointer_info`. 

Only, `dev_runtime_unloaded_` keeps the target agnostic shutdown state at `DeviceAllocator`, avoiding this check here is not a good option(sorry based on my thinking) . @bing-ma  to comment more?

b-sumner replied:
I appreciate your comments, but I'm hoping for some more specific analysis than "chances of being in an inconsistent state".  I'm wondering what specifically is problematic with the code that the return skips which

- removes the chunk of interest and updates the stats
- executes the unmap callback
- Calls DeviceMemFuncs::Deallocate

ampandey-AMD replied:
Ok, I was thinking under the assumption that Device ASan Allocator state about the Device memory(chunks allocated/deallocated/tracked) via `DeviceAllocator` would get corrupt due to fallback to single page and removing this check would also unposion(unmap callback) the shadow memory of that single page but actual request was something(large than 1 page usually) else but that was wrong assumption.

I figured out now that `GetHeader` at line 124 would actually fallback return to a single page post shutdown and then doing unmap callback. So, we are actually calling into DeviceMemFuncs::Deallocate after stats manipulations with single page post-shutdown which ulimately is no-OP.

Resolved with commit [d57b151](https://github.com/ROCm/llvm-project/pull/559/commits/d57b15191d08d037813ec090b38300533696e1ca)

bing-ma replied:
> @bing-ma to comment more?
Hi @ampandey-AMD, sorry for the late reply. Next time ping me in MS teams if you need quick response. I see this is already resolved. Let me know if you need further comments.

b-sumner commented on compiler-rt/lib/sanitizer_common/sanitizer_allocator_device.h:
```diff
@@ -140,7 +141,8 @@ class DeviceAllocatorT {
     for (uptr i = 0; i < n_chunks_; i++) {
       Header *h = GetHeader(chunks_[i], &header);
-      CHECK(!dev_runtime_unloaded_);
+      if (UNLIKELY(dev_runtime_unloaded_))
```
I don't think we need to give up here now either.

ampandey-AMD replied:
Same here, only at two places in `DeviceAllocator` layer we are early checking `dev_runtime_unloaded_` , one at Deallocate and other at `TotalMemoryUsed`  removing these would cause inconistency in sanitizer asan_allocator layer which only has knowledge about `DeviceAllocator` that tracks and maintains these device allocations.

b-sumner replied:
Thanks but again, I'm hoping for a more specific analysis.  What is problematic about the remaining code here which updates "res" and continues iterating?

ampandey-AMD replied:
Resolved with commit [d57b151](https://github.com/ROCm/llvm-project/pull/559/commits/d57b15191d08d037813ec090b38300533696e1ca)

ampandey-AMD committed:
Address @b-sumner comments

Remove early checks for functions Deallocate & TotalMemoryUsed in
sanitizer_allocator_device.h.Early return at this place is not required
since `GetHeader` acts as a guard and postshutdown only single page
operations are issued to actual actual Deallocate & TotalMemoryUsed
which are no-op at `AmdgpuAllocator`.

z1-cciauto commented:
PSDB Link: https://compiler-ci.amd.com/job/compiler-psdb-amd-staging/3298

ampandey-AMD commented:
Hi @b-sumner ,
Should I add positive/negative reporting for `Allocation/Deallocation/GetPointerInfo`. 
I don't think we should add more `VReport`.
Any Comments?

b-sumner commented:
> Hi @b-sumner , Should I add positive/negative reporting for `Allocation/Deallocation/GetPointerInfo`. I don't think we should add more `VReport`. Any Comments?

I agree, I don't think reporting in these is needed at this time.

b-sumner reviewed (approved):
LGTM.  Thanks!

bing-ma reviewed (approved):
LGTM

ampandey-AMD commented:
Thanks all for the approval.
